resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala

```diff
@@ -67,8 +66,16 @@ private[spark] object SparkKubernetesClientFactory {
     val dispatcher = new Dispatcher(
       ThreadUtils.newDaemonCachedThreadPool("kubernetes-dispatcher"))
 
-    // TODO [SPARK-25887] Create builder in a way that respects configurable context
-    val config = new ConfigBuilder()
+    // Allow for specifying a context used to auto-configure from the users K8S config file
+    val kubeContext = sparkConf.get(KUBERNETES_CONTEXT).filter(c => StringUtils.isNotBlank(c))
+    logInfo(s"Auto-configuring K8S client using " +
+      s"${if (kubeContext.isEmpty) s"context ${kubeContext.get}" else "current context"}" +
+      s" from users K8S config file")
+
+    // Start from an auto-configured config with the desired context
+    // Fabric 8 uses null to indicate that the users current context should be used so if no
+    // explicit setting pass null
+    val config = new ConfigBuilder(autoConfigure(kubeContext.getOrElse(null)))
```

Comment:
What happens here when the context does not exist? Does it fall back to the default?

e.g. in cluster mode, the config you're adding will be propagated to the driver, and then this code will be called with the same context as the submission node. What if that context does not exist inside the driver container?

Reply:
If the context does not exist then Fabric 8 falls back to other ways of auto-configuring itself (e.g. service account)

Fabric 8 skips any file based auto-configuration if there is no K8S config file present (https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L436-L459).

Since we don't propagate the submission clients config file into the driver pods no auto-configuration from config file will be attempted in the driver because there won't be a config file present.

Reply:
> we don't propagate the submission clients config file

We don't propagate the config *file* itself, but we do propagate all its contents, as far as I remember. But given your explanation it should work fine, unless there's a different config context with the same name inside the container...

Reply:
(Just to clarify I'm referring to the Spark config. Too many config files involved here...)

Reply:
Yes and I was referring to the K8S config file :) And yes the fact that we would propagate `spark.kubernetes.context` into the pod shouldn't be an issue because there won't be any K8S config file for it to interact with inside the pod as in-pod K8S config should be from the service account token that gets injected into the pod

Reply:
This somewhat answers my previous question, and it makes sense. Seems like Fabric8 is behaving consistently with other SDKs (e.g. AWS comes to mind as it will try 3-4 default configuration options, including env vars, iam machine role, etc).

That being said, if you were to deploy an app inside a docker container, it won't see any of those and one still needs to inject some env vars or config maps for the AWS SDK to pick up.

Long story short, I think this enhancement does not apply to client mode, and we should make that clearer in the docs, while providing alternatives for issues like https://issues.apache.org/jira/browse/SPARK-26295.

Hopefully I'm making sense.

Reply:
> I think this enhancement does not apply to client mode

If you mean "client mode inside a k8s-managed docker container", then yes, you may need to do extra stuff, like mount the appropriate credentials. But in the "client mode with driver inside k8s pod" case, Spark does not create that pod for you. So I'm not sure how Spark can help with anything there; the `serviceName` configuration seems targeted at propagating the credentials of the submitter to the driver pod, and in that case Spark is not creating the driver pod at all.

Reply:
Ok, this is starting to sink in. I just realized that in my deployment scenario, all I need to do is to specify the `serviceAccountName` on _my_ pod template that contains the driver.

That leaves me confused though - what other "client mode deployments" do you see happening on K8S? What does client mode mean to you?
Also - how should one interpret this paragraph in the docs?
` In client mode, use spark.kubernetes.authenticate.serviceAccountName instead.`

Reply:
> What does client mode mean to you?

Client mode means that the driver is not started by Spark. It's started directly by the user.

> Also - how should one interpret this paragraph in the docs?

I have no idea.

Reply:
Of course :) I think what I'm asking is what real-world options do we forsee for client mode in a k8s context. I don't see many beyond a pod that the user creates. Whether straight spark or something like jupyter/zeppelin notebook, at the end of the day it's probably a docker container living in a pod.

Thanks for taking the time, I won't spam this thread anymore. I'll continue the discussion on the issue I created, which I'll repurpose for improving the docs in this area.